src/ceph-volume/ceph_volume/util/disk.py

```diff
@@ -195,6 +195,26 @@ def device_family(device):
 
     return devices
 
+def is_rotational(device):
+    """
+    Returns true if a device is rotational or not.
+    """
+    labels = ['ROTA']
+    command = ['lsblk', '-P', '-p', '-o', ','.join(labels), device]
+    out, err, rc = process.call(command)
+    device = []
+    for line in out:
+        device.append(_lsblk_parser(line))
+
+    for i in device:
+        if type(i) is dict and "ROTA" in i.keys():
+            if i['ROTA'] == '1':
+                rotational = True
+            else:
+                rotational = False
+
+    return rotational
+
```

Comment:
> `src/common/blkdev.cc` and `src/ceph-volume/ceph_volume/util/device.py` and `src/pybind/mgr/rook/rook_cluster.py` have existing functions that serve this purpose. Is this wheel reinvention?

Note that  `src/ceph-volume/ceph_volume/util/device.py`  includes `disk.py`. AFAICT it's not duplicating any effort here.

I have taken a look at `src/pybind/mgr/rook/rook_cluster.py` , but it does not look like it's performing similar functions as `disk.py`.

I agree that a common library, shared between rook/cephadm/ceph-volume, might make sense. But that seems to be something that has to be discussed between maintainers of those projects

> Are we doing it the same way across the code?

Yes, this is how it's done for other disk properties as well.
 
> Might inspecting `/sys/block` be cleaner than shelling out to `lsblk`? Can we count on `lsblk` being always installed and in the path?

If `lsblk` is doing a good job Ceph does not have to reinvent the wheel here. It's used in ceph-volume in lots of places. I have not checked what distro's or containers come with lsblk by default. It should be a dependency of ceph-osd. I checked on ubuntu, and it's not listed as a dependency there. Containers (created by Ceph) will have this command available.  I do want to stress however, that the use of `lsblk` here does not add any new dependencies, it's just another part of the code that makes use of it. And that the general question if that is a good thing or not should be discussed on a higher level than this PR.

Reply:
I wonder how each implementation behaves on top of LVM, dmcrypt, and when containerized.  I would not be surprised if there are clusters out there crippled by specious determinations of rotation.

Reply:
> This is a good catch, Anthony.
> 
> @hydro-b It looks like there are now three ways in ceph-volume that the rotational state of a drive is inspected; the first two pre-existed your changes:
> 
>     1. A property returned from disk.py's `get_devices()`
> 
>     2. device.py's `Device.rotational` (this seems like the most complete implementation)
> 
>     3. The new `is_rotational()` added here
> 
> 
> Do you think there's a way to unify the implementations of 2 and 3, or, per the other comment, at least 1 and 3? At a glance I'm not sure how this would be done without a few rounds of refactoring.

I've looked at using 1) and 2).

1) disk.py's `get_devices()` excludes logical volumes
2) device.py's rejects the LVM volume it gets fed. It ends with a general rejected reason. IIRC the issue has to do with /dev/vg/lv being a symlink to a device mapper device, the format of the symlink: `/dev/vg/lv` instead of `vg.lv`. I've hacked up a `lsblk` version does a regex match and fixes the device into what lsblk wants. And then my time to work on this was up and had to shift to operational stuff.

So yeah, I agree that a single function to return rotational state for whatever device type / format gets passed would be nice, however it's not there yet.